.github: Add label generation tool and update labels

arvganesh opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

Added `update_labels.py`, a tool to automatically update the files in `.github/labels` with configuration of latest labels from Github.

It only has one external dependency, [ruamel-yaml](https://yaml.readthedocs.io/en/latest/), which is a more powerful derivative of [PyYAML](https://pypi.org/project/PyYAML/#:~:text=YAML%20is%20a%20data%20serialization,parser%20and%20emitter%20for%20Python.) for interacting with YAML files.

### Description of changes

Update labels in `.github/labels` to match the current state of the repo.

To aid with this process in the future, I also wrote a tool (`update_labels.py`) to automate the majority of this process. It queries [this endpoint](https://docs.github.com/en/rest/issues/labels?apiVersion=2022-11-28#list-labels-for-a-repository) in the GithubAPI, grabs the relevant data, and updates the YAML files in `.github/labels`.

GitHub-Fixes: #1283

craciunoiuc commented on .github/update_labels.py:
```diff
@@ -0,0 +1,135 @@
+# Usage: python update_labels.py
+if __name__ == "__main__":
+    labels = get_labels()
+    update_repo(labels)
```
```suggestion
    update_repo(labels)

```

Missing newline at the end

craciunoiuc commented on .github/update_labels.py:
```diff
@@ -0,0 +1,135 @@
+        labels_from_file = yaml_context.load(old_yaml_file)
+
+    output_file = open(label_path, "w")
+    
```
Some trailing whitespace here

craciunoiuc commented on .github/update_labels.py:
```diff
@@ -0,0 +1,135 @@
+from collections import defaultdict
```
Can you move this script and the template to a separate `scripts/` directory?

craciunoiuc reviewed (changes requested):
Hey @arvganesh 

This initial version looks really good, left you some comments.

Can you squash you commits and leave only two:
1. `.github/scripts: Add label generation tool`
2. `.github/labels: Update labels to track latest staging`

You will also need to add commit message bodies for each to describe the changes. (in order for the `checkpatch` check to pass)

craciunoiuc commented:
In a separate PR, if you want, we can also add a workflow that runs on pushes (or weekly) and checks if labels changed and updates them automatically.

arvganesh committed:
.github/labels: Update labels to track latest staging

Using the code generation tool in .github/scripts, update labels in .github/labels to match the latest staging.

Signed-off-by: Arvind Ganesh <arvind.cganesh@gmail.com>

nderjung reviewed (changes requested):
Hi @arvganesh! Thanks for this contribution. We use an external tool called [`governctl`](https://github.com/unikraft/governance) to do the heavy lifting in terms of interacting with GitHub's API, performing operations specifically on the Unikraft core repository as well as organization-wide repositories and teams.

In order to prevent too many tools, scripts or programs dotted all over the place, I suggest looking at how label synchronization can be accomplished with this tool under a new subcommand `governctl labels sync`.

One thing to note is that these label config files are in fact meant to work in the opposite direction: their static declaration should dictate GitHub's state of labels (aka "Infrastructure as Code").  This means that labels should be updated from the repo and not the other way around.

I think this particular issue can be kept nearly as-is sans the introduction of the python script, changes to descriptions and changing the hex color values. 

For direct changes to the labels, please:

- Ensure all color attributes are quoted, lowercase and include a pound (`#`) symbol i.e.: `"#1d76db"` is better vs. `1D76DB`.  This prevents the diff from needless changes;
- Only change the description if it is a better representation of what the label is intended;
- Remove `libx/posix-unisocket`;

Thanks!

arvganesh commented:
Hi @nderjung. Thanks for getting back to me! Had a few follow-up questions:

Re: `governctl labels sync`, are you suggesting that I should look into adding the label generation code to `governctl`? 

Re: The changes in this PR, there were many label files that were "created" (i.e labels were on GitHub but not in `./github/labels`). Since you said the contents of `.github/labels` are the ground truth, should I revert any new label files that were created?

nderjung commented:
> Re: governctl labels sync, are you suggesting that I should look into adding the label generation code to governctl?

Yes.

> Re: The changes in this PR, there were many label files that were "created" (i.e labels were on GitHub but not in ./github/labels). Since you said the contents of .github/labels are the ground truth, should I revert any new label files that were created?

The new libraries we can keep.  There are also changes which need to be reverted:

- There are additions and deletions which represent re-ordering of libraries.  E.g. `bug/fix`, `lifecycle/stale`, `lifecycle/rotten`, `priority/low`, etc.  Please double check all entries and please keep their order.
- A lot of descriptions are changed, I assume because they were taken what GitHub has.  Having gone through the list, I think the best approach is to take the longer description.  E.g. "Issues and PRs related to Linux Userspace platform" vs "Unikraft for Linux Userspace", the former is better.

Thanks!

razvand commented:
@arvganesh , could you please see @nderjung 's comments?
